From 0070ac88f56ada9a9dc2877eaea82c690fb9e904 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Wed, 17 Aug 2022 08:50:45 -0700
Subject: [PATCH] [docs] Add dependencies and build instructions for NNG and
 flatbuffers (#265)

Add in missing documentation about flatbuffers and nng dependencies to
the build documentation.

Co-authored-by: PiRK <pierrekn@gmail.com>

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: 006ff2cc99ae6b3b55d983c464c368c8fc73e375
Recorded-replay-commit: d94593ec8093f33a05297889e35bad4ea913e0b8

diff --git a/doc/build-unix.md b/doc/build-unix.md
index 01bc071e3a..1c5c8b0626 100644
--- a/doc/build-unix.md
+++ b/doc/build-unix.md
@@ -44,6 +44,8 @@ Optional dependencies:
  libqrencode | QR codes in GUI  | Optional for generating QR codes (only needed when GUI enabled)
  univalue    | Utility          | JSON parsing and encoding (bundled version will be used unless --with-system-univalue passed to configure)
  libzmq3     | ZMQ notification | Optional, allows generating ZMQ notifications (requires ZMQ version >= 4.1.5)
+ nng         | NNG interface    | See [nng.md](nng.md)
+ flatbuffers | Serialization    | Used by the NNG interface
  systemtap   | Tracing (USDT)   | Optional, statically defined tracepoints
 
 For the versions used, see [dependencies.md](dependencies.md)
@@ -114,6 +116,27 @@ jemalloc dependencies (provides the jemalloc library, can be disabled by passing
 
     sudo apt-get install libjemalloc-dev
 
+nng dependencies (can be disabled by passing `-DBUILD_BITCOIN_NNG=OFF` on the cmake command line):
+
+    sudo apt-get install libnng-dev
+
+FlatBuffers (if 2.0.0 is available in the package repository):
+
+    sudo apt-get install libflatbuffers-dev
+
+If the repository does not have the correct version, install it from source:
+
+    wget https://github.com/google/flatbuffers/archive/refs/tags/v2.0.0.tar.gz
+    echo "9ddb9031798f4f8754d00fca2f1a68ecf9d0f83dfac7239af1311e4fd9a565c4 v2.0.0.tar.gz" | sha256sum -c
+    tar -zxf v2.0.0.tar.gz
+    (
+      cd flatbuffers-2.0.0
+      mkdir build
+      cd build
+      cmake -GNinja -DCMAKE_BUILD_TYPE=Release ..
+      sudo ninja install
+    )
+
 User-Space, Statically Defined Tracing (USDT) dependencies (provides the tracepoint library, can be disabled by passing `-DENABLE_TRACING=OFF` on the cmake command line):
 
     sudo apt install systemtap-sdt-dev
diff --git a/doc/dependencies.md b/doc/dependencies.md
index f3d9d3b277..e462640828 100644
--- a/doc/dependencies.md
+++ b/doc/dependencies.md
@@ -9,6 +9,7 @@ These are the dependencies currently used by Bitcoin ABC. You can find instructi
 | Boost | [1.81.0](https://www.boost.org/users/download/) | 1.64.0 | No |  |  |
 | Clang |  | [10](https://releases.llvm.org/download.html) (C++20 support) |  |  |  |
 | CMake |  | [3.25](https://cmake.org/download/) |  |  |  |
+| FlatBuffers |  | [2.0.0](https://github.com/google/flatbuffers/releases) |  |  |  |
 | fontconfig | [2.12.6](https://www.freedesktop.org/software/fontconfig/release/) |  | No | Yes |  |
 | FreeType | [2.11.0](http://download.savannah.gnu.org/releases/freetype) |  | No |  |  |
 | GCC |  | [12](https://gcc.gnu.org/) |  |  |  |
@@ -19,6 +20,7 @@ These are the dependencies currently used by Bitcoin ABC. You can find instructi
 | libpng |  |  |  |  | Yes |
 | librsvg | |  |  |  |  |
 | Ninja |  | [1.5.1](https://github.com/ninja-build/ninja/releases) |  |  |  |
+| nng |  | [1.5.2](https://github.com/nanomsg/nng/releases) |  |  |  |
 | OpenSSL | [1.0.1k](https://www.openssl.org/source) |  | Yes |  |  |
 | PCRE |  |  |  |  | Yes |
 | protobuf | [21.12](https://github.com/protocolbuffers/protobuf/releases/tag/v21.12) |  | No |  |  |
@@ -45,3 +47,4 @@ factors that affect the dependency list.
 * qrencode is not needed with `-DENABLE_QRCODE=OFF`.
 * systemtap is not needed with `-DENABLE_TRACING=OFF`.
 * ZeroMQ is not needed with the `-DBUILD_ZMQ=OFF`.
+* nng and FlatBuffers are not needed with the `-DBUILD_BITCOIN_NNG=OFF`.
